### PR TITLE
[dashboard] customer notes backend

### PR DESCRIPTION
**Pull request by fkotsian**
#### Schema
```
Note:
- id: int
- author_id: int
- account_id: uuid
- customer_id: uuid
- body: text
- timestamps()
```

### Issue

Part of https://github.com/papercups-io/papercups/issues/433

### Screenshots

N/A

## Checklist

- [X] Everything passes when running `mix test`
- [X] Ran `mix format`
- [X] No frontend compilation warnings

#### Current UX:
N/A

### CODE REVIEW

**Comment by fkotsian** [comment]
@reichert621 Here's a first stab at the functionality for the CustomerNotes feature. As you said I think a lot of work is on the FE/UX we want -- thought we could approve/sync on this and then polish the UX on Friday?

#### Open Qs in this PR

- I am wondering if the account_id key should be :through customer (since customer belongs_to an :account? Or how should we fetch/pass the account in to the ConvDetailsSidebar)
- # of notes to display? (suggest: 3?)
- note 'show' UI? (suggest: bordered box, as in Intercom?)
- error state (suggest: border highlight?)
- delete ux? (suggest: 'x'?)
- edit ux? (suggest: none?)
- should call the table Notes.CustomerNote customer_notes as opp to Notes.Note (future flexibility?)

#### UXs needed:

- delete (trigger? modal? confirmation?)
- edit (suggest: no edit at this time?)
- validate/error (suggest: a simple border highlight - Papercups red? no tooltip - only poss error state is "no input"?)
- any others?

My suggestion would be to go with the most minimal experience for each of these and refine later? Whatever is best for the product of course 🙌

File: test/chat_api/notes_test.exs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  alias ChatApi.Notes
+
+  describe "notes" do
+    alias ChatApi.Notes.Note
+
+    @valid_attrs %{body: "some body"}
+    @update_attrs %{body: "some updated body"}
+    @invalid_attrs %{body: nil}
+
+    def note_fixture(attrs \\ %{}) do
```
Comment (flmel) [inline test/chat_api/notes_test.exs:13]: We have moved to factories from fixtures. Have a look on factory.ex you will need to create something like 

```
def note_factory do 
    %ChatApi.Notes.Note{ 
          account: build(:account),
          author: build(:user),
          customer: build(:customer)
     }
end
```
You will need to import ChatApi.Factory and then within your tests you will have access to ex_machinas comands
insert,insert_pair,insert_list, build, params_for, params_with_assocs for more info you can see [ex_machina docs](https://hexdocs.pm/ex_machina/readme.html) or peek into accounts_test.ex (or other tests) to see it in use.

Let me know if I can be of any help.

File: lib/chat_api/notes/note.ex [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    field :customer_id, :binary_id
+
+    timestamps()
+  end
+
+  @doc false
+  def changeset(note, attrs) do
+    note
+    |> cast(attrs, [:body, :author_id, :customer_id])
+    |> validate_required([:body, :author_id, :customer_id])
```
Comment (reichert621) [inline lib/chat_api/notes/note.ex:20]: don't forget the `account_id` on the two lines above!

File: lib/chat_api/notes/note.ex [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  import Ecto.Changeset
+
+  @primary_key {:id, :binary_id, autogenerate: true}
+  @foreign_key_type :binary_id
+  schema "notes" do
+    field :body, :string
+    field :author_id, :integer
+    field :account_id, :binary_id
+    field :customer_id, :binary_id
+
```
Comment (reichert621) [inline lib/chat_api/notes/note.ex:12]: can you set up the `belongs_to` relations here too? e.g.

```elixir
    belongs_to(:account, Account)
    belongs_to(:customer, Customer)
    belongs_to(:author, User, foreign_key: :author_id, references: :id, type: :integer)
```

File: priv/repo/migrations/20201210043230_create_notes.exs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+defmodule ChatApi.Repo.Migrations.CreateNotes do
+  use Ecto.Migration
+
+  def change do
+    create table(:notes, primary_key: false) do
+      add :id, :binary_id, primary_key: true
+      add :body, :text
+      add :author_id, references(:users, on_delete: :nothing, type: :id)
+      add :account_id, references(:accounts, on_delete: :nothing, type: :binary_id)
```
Comment (reichert621) [inline priv/repo/migrations/20201210043230_create_notes.exs:9]: let's do `on_delete: :delete_all` here -- if an account is deleted, we want all the associated notes to be deleted as well

File: priv/repo/migrations/20201210043230_create_notes.exs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      add :customer_id, references(:customers, on_delete: :nothing, type: :binary_id)
```
Comment (reichert621) [inline priv/repo/migrations/20201210043230_create_notes.exs:10]: likewise here -- if a customer is deleted, all its associated notes should be deleted as well

File: test/chat_api/notes_test.exs (hunk 1 above)
Comment (reichert621) [reply test/chat_api/notes_test.exs:13]: i think `author` is fine, since it's a bit more descriptive than `user` (Intercom uses `author` as well in their notes API: https://developers.intercom.com/intercom-api-reference/reference#note-model -- not that we should blindly copy them, but i think it's a reasonable name 😄 )

File: lib/chat_api_web/router.ex [hunk 6]
```diff
@@ -98,7 +98,11 @@ defmodule ChatApiWeb.Router do
     resources("/accounts", AccountController, only: [:update, :delete])
     resources("/messages", MessageController, except: [:new, :edit])
     resources("/conversations", ConversationController, except: [:new, :edit, :create])
-    resources("/customers", CustomerController, except: [:new, :edit, :create])
+
+    resources("/customers", CustomerController, except: [:new, :edit, :create]) do
+      resources("/notes", NoteController, except: [:new, :edit])
```
Comment (reichert621) [inline lib/chat_api_web/router.ex:103]: i'd prefer not to nest this resource -- in general, i think it's best to keep as many resources at the top level (i.e. `/api/notes` if possible. (this is how we did it at stripe, and i think this stack overflow comment sums up my feelings nicely: https://stackoverflow.com/a/36410780)

File: lib/chat_api_web/controllers/note_controller.ex [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  use ChatApiWeb, :controller
+
+  alias ChatApi.Notes
+  alias ChatApi.Notes.Note
+
+  action_fallback ChatApiWeb.FallbackController
+
+  @spec index(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def index(conn, _params) do
+    notes = Notes.list_notes()
```
Comment (reichert621) [inline lib/chat_api_web/controllers/note_controller.ex:11]: this method currently fetches all of the notes in the entire database -- let's make sure we filter by `customer_id` and/or `account_id`

(remember, we can get the `account_id` from `conn.assigns.current_user.account_id`)

File: lib/chat_api_web/controllers/note_controller.ex [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    render(conn, "index.json", notes: notes)
+  end
+
+  @spec create(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def create(conn, %{"note" => note_params}) do
+    with {:ok, %Note{} = note} <- Notes.create_note(note_params) do
```
Comment (reichert621) [inline lib/chat_api_web/controllers/note_controller.ex:17]: let's make sure when we're creating a new `note`, we don't allow the client to specify the `account_id` or the `author_id` -- these should come directly from `conn.assigns.current_user`

File: lib/chat_api_web/views/note_view.ex [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+  def render("show.json", %{note: note}) do
+    %{data: render_one(note, NoteView, "note.json")}
+  end
+
+  def render("note.json", %{note: note}) do
+    %{
+      id: note.id,
+      body: note.body
```
Comment (reichert621) [inline lib/chat_api_web/views/note_view.ex:16]: i think it's fine to include the `customer_id` and `author_id` here too, will probably help with debugging

File: test/chat_api_web/controllers/note_controller_test.exs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        get(conn, Routes.note_path(conn, :show, note))
+      end
+    end
+  end
+
+  defp create_note(_) do
+    note = fixture(:note)
+    %{note: note}
+  end
+end
```
Comment (reichert621) [inline test/chat_api_web/controllers/note_controller_test.exs:88]: let's get all the backend stuff fixed with tests passing before we worry about the UX :)

**Review by reichert621 — CHANGES_REQUESTED** [review]
hey @fkotsian -- nice work so far! 😄 

i mainly focused on reviewing the backend code to start off... it would be great to get all the backend stuff fixed up with tests passing and then merge that in a separate PR before we worry about the UX 👍

File: test/chat_api/notes_test.exs (hunk 1 above)
Comment (reichert621) [reply test/chat_api/notes_test.exs:13]: i think `author` is fine, since it's a bit more descriptive than `user` (Intercom uses `author` as well in their notes API: https://developers.intercom.com/intercom-api-reference/reference#note-model -- not that we should blindly copy them, but i think it's a reasonable name 😄 )

File: lib/chat_api/notes/note.ex (hunk 3 above)
Comment (reichert621) [reply lib/chat_api/notes/note.ex:12]: (might be nice to add the `has_many` relation to the `ChatApi.Customers.Customer` module in `customer.ex` as well)

File: lib/chat_api_web/controllers/note_controller.ex (hunk 7 above)
Comment (fkotsian) [reply lib/chat_api_web/controllers/note_controller.ex:11]: Great, this clears up my main question

File: lib/chat_api_web/router.ex (hunk 6 above)
Comment (fkotsian) [reply lib/chat_api_web/router.ex:103]: V nice! People def have differing opinions about this

File: lib/chat_api/notes/note.ex (hunk 2 above)
Comment (fkotsian) [reply lib/chat_api/notes/note.ex:20]: Nice

File: lib/chat_api_web/controllers/note_controller.ex (hunk 7 above)
Comment (fkotsian) [reply lib/chat_api_web/controllers/note_controller.ex:11]: Nice call!

File: lib/chat_api_web/router.ex (hunk 6 above)
Comment (fkotsian) [reply lib/chat_api_web/router.ex:103]: Works for me!

File: test/chat_api/notes_test.exs (hunk 1 above)
Comment (fkotsian) [reply test/chat_api/notes_test.exs]: Thanks @flmel , the factories are much nicer to use than fixtures! I gave it a shot - let me know what you think?

File: lib/chat_api/notes/note.ex [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    belongs_to(:account, Account)
+    belongs_to(:customer, Customer)
+    belongs_to(:author, User, foreign_key: :author_id, references: :id, type: :integer)
+
+    timestamps()
+  end
+
+  @doc false
+  def create_changeset(note, attrs) do
```
Comment (reichert621) [inline lib/chat_api/notes/note.ex:22]: i think we can just use `changeset` below, this `create_changeset` shouldn't be necessary 👍 (reference: https://hexdocs.pm/ecto/Ecto.Changeset.html#validate_required/3)

> If a field is given to validate_required/3 but it has not been passed as parameter during cast/3 (i.e. it has not been changed), then validate_required/3 will check for its current value in the data. If the data contains a non-empty value for the field, then no error is added. **This allows developers to use validate_required/3 to perform partial updates. For example, on insert all fields would be required, because their default values on the data are all nil, but on update, if you don't want to change a field that has been previously set, you are not required to pass it as a parameter**, since validate_required/3 won't add an error for missing changes as long as the value in the data given to the changeset is not empty.

File: priv/repo/migrations/20201210043230_create_notes.exs [hunk 12]
```diff
@@ -0,0 +1,19 @@
+defmodule ChatApi.Repo.Migrations.CreateNotes do
+  use Ecto.Migration
+
+  def change do
+    create table(:notes, primary_key: false) do
+      add :id, :binary_id, primary_key: true
+      add :body, :text
```
Comment (reichert621) [inline priv/repo/migrations/20201210043230_create_notes.exs:7]: sorry i missed this earlier -- let's add `null: false` to all the required fields, e.g.
```
add :body, :text, null: false
```

File: test/chat_api_web/controllers/note_controller_test.exs [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    } do
+      conn =
+        get(authed_conn, Routes.note_path(authed_conn, :index), %{
+          "note" => %{customer_id: note.customer_id}
+        })
+
+      note_ids = json_response(conn, 200)["data"] |> Enum.map(& &1["id"])
+
+      assert note_ids == [note.id]
+    end
```
Comment (reichert621) [inline test/chat_api_web/controllers/note_controller_test.exs:43]: can we also add a test to see what happens when a `customer_id` param is missing from the request?

(ideally, it would return a `400` error)

File: test/chat_api_web/controllers/note_controller_test.exs [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+      assert %{
+               "id" => created_id,
+               "customer_id" => customer_id,
+               "author_id" => agent_id
+             } = json_response(conn, 201)["data"]
+
+      # note is fetchable
+      conn = get(authed_conn, Routes.note_path(authed_conn, :show, created_id))
+      _fetched = json_response(conn, 200)["data"]
```
Comment (reichert621) [inline test/chat_api_web/controllers/note_controller_test.exs:69]: just curious, why `_fetched` here instead of just `fetched`?

File: test/chat_api_web/controllers/note_controller_test.exs [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+    } do
+      # update note
+      conn = put(authed_conn, Routes.note_path(authed_conn, :update, note), note: @update_attrs)
+      assert %{"id" => ^id} = json_response(conn, 200)["data"]
+
+      # fetch updated note
+      conn = get(authed_conn, Routes.note_path(conn, :show, id))
+
+      assert %{
+               "id" => _id,
```
Comment (reichert621) [inline test/chat_api_web/controllers/note_controller_test.exs:100]: i think we can use `^id` here as well to make sure the id still matches above?

File: test/chat_api/notes_test.exs [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    setup do
+      account = insert(:account)
+      agent = insert(:user, account: account)
+      customer = insert(:customer, account: account)
+      note = insert(:note, body: "note_body", account: account, author: agent, customer: customer)
+
+      {:ok, agent: agent, account: account, customer: customer, note: note}
+    end
+
+    test "list_notes/1 returns all notes for the given account and customer", %{
```
Comment (reichert621) [inline test/chat_api/notes_test.exs:22]: typo -- `list_notes` -> `list_notes_for_customer`

File: lib/chat_api_web/controllers/note_controller.ex [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+      render(conn, "index.json", notes: notes)
+    end
+  end
+
+  @spec create(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def create(conn, %{"note" => note_params}) do
+    %{"body" => body, "customer_id" => customer_id} = note_params
+
+    with %{account_id: account_id, id: author_id} <- conn.assigns.current_user do
+      with {:ok, %Note{} = note} <-
```
Comment (reichert621) [inline lib/chat_api_web/controllers/note_controller.ex:24]: no need to nest `with` statements :) we can just do:
```elixir
with %{account_id: account_id, id: author_id} <- conn.assigns.current_user,
  {:ok, %Note{} = note} <- Notes.create_note(%{...}) do
  # etc
end
```

(some docs on `with`: https://elixirschool.com/en/lessons/basics/control-structures/#with)

File: lib/chat_api_web/controllers/note_controller.ex [hunk 18]
```diff
@@ -0,0 +1,65 @@
+defmodule ChatApiWeb.NoteController do
+  use ChatApiWeb, :controller
+
+  alias ChatApi.Notes
+  alias ChatApi.Notes.Note
+
+  action_fallback ChatApiWeb.FallbackController
+
+  @spec index(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def index(conn, %{"note" => note_params}) do
```
Comment (reichert621) [inline lib/chat_api_web/controllers/note_controller.ex:10]: for the params, i think we can simplify it to just `%{"customer_id" => customer_id}` (rather than `%{"note" => "customer_id" => customer_id}}`)

then the API request would just look like
```
GET /api/notes?customer_id=xxxxx
```

File: test/chat_api/notes_test.exs [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+                 author_id: agent.id
+               })
+
+      assert note.body == "some body"
+      assert note.customer_id == customer.id
+      assert note.author_id == agent.id
+      assert note.account_id == account.id
+    end
+
+    test "create_note/1 with invalid data returns error changeset" do
```
Comment (reichert621) [inline test/chat_api/notes_test.exs:59]: can we also test that attempting to create a note with an empty string body also returns an error changeset?

**Review by reichert621 — CHANGES_REQUESTED** [review]
nice! backend stuff looks great, just a few minor things and then it should be good to go 👍

File: lib/chat_api/notes/note.ex (hunk 11 above)
Comment (fkotsian) [reply lib/chat_api/notes/note.ex:22]: Updated! My thought here was actually that allowing the entire changeset in update requests allows the client to change the Customer or the Account that the note displays on -- can add back if we're concerned about this!

File: test/chat_api_web/controllers/note_controller_test.exs (hunk 14 above)
Comment (fkotsian) [reply test/chat_api_web/controllers/note_controller_test.exs:69]: ooh good catch, test shoulda been an `==`

File: test/chat_api_web/controllers/note_controller_test.exs (hunk 15 above)
Comment (fkotsian) [reply test/chat_api_web/controllers/note_controller_test.exs:100]: These pins got me... I think I got 'em now. 

- pin when var already exists
- otherwise gets interpreted as reassignment

File: lib/chat_api_web/controllers/note_controller.ex (hunk 17 above)
Comment (fkotsian) [reply lib/chat_api_web/controllers/note_controller.ex:24]: Ah: with is a context (return context), not an extraction operator (that's just the `<-` part) 💡

Comment (fkotsian) [reply lib/chat_api_web/controllers/note_controller.ex:24]: Huh, it took me a while but I think I get this..

So
```
with %{account_id: account_id, id: author_id} <- conn.assigns.current_user,
  {:ok, %Note{} = note} <- Notes.create_note(%{...}) do
```

if the {:ok, %Note{}} doesn't match (for ex, if create_note returns an {:error, Changeset}), it just falls through to outside the `with` (and in this case is handled by our FallbackController)?

**Comment by fkotsian** [comment]
@reichert621 reverted all the frontend diff for ya, should be good to go!

**Review by reichert621 — APPROVED** [review]
woohoo looks great! thanks @fkotsian 🎉🎉🎉
